cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -404,174 +409,231 @@ func (m *McmManager) GetMachineDeploymentSize(machinedeployment *MachineDeployme
-			} else if err != nil {
-				// Timeout occurred
-				klog.Errorf("Unable to fetch Machine object %s, Error: %s", machine.Name, err)
-				return err
+// resetPriorityForNotToBeDeletedMachines resets the priority of machines with nodes without ToBeDeleted taint to 3
+func (m *McmManager) resetPriorityForNotToBeDeletedMachines(mdName string) error {
+	allMachinesForMachineDeployment, err := m.getMachinesForMachineDeployment(mdName)
+	if err != nil {
+		return fmt.Errorf("unable to list all machines for node group %s, cannot proceed with scale-in of machines, Error: %+v", mdName, err)
+	}
+
+	for _, machine := range allMachinesForMachineDeployment {
+		nodeName := machine.Labels[v1alpha1.NodeLabelKey]
+		err = func() error {
+			node, err := m.nodeLister.Get(nodeName)
+			if err != nil && !kube_errors.IsNotFound(err) {
+				return fmt.Errorf("unable to get Node object %s for machine %s, cannot proceed with scale-in of machines, Error: %+v", nodeName, machine.Name, err)
+			} else if err == nil && taints.HasToBeDeletedTaint(node) {
+				// Don't update priority annotation if the taint is present on the node
+				return nil
 			}
-
-			mclone := machine.DeepCopy()
-
-			if isMachineTerminating(mclone) {
-				terminatingMachines = append(terminatingMachines, mclone)
-			} else {
-				expectedToTerminateMachineNodePairs = append(expectedToTerminateMachineNodePairs, machineNameNodeNamePair{mclone.Name, mclone.Labels["node"]})
+			val, ok := machine.Annotations[machinePriorityAnnotation]
+			if ok && val != "3" {
+				ctx, cancelFn := context.WithDeadline(context.Background(), time.Now().Add(m.maxRetryTimeout))
+				defer cancelFn()
+				for {
+					mc, err := m.machineLister.Machines(m.namespace).Get(machine.Name)
+					if err != nil {
+						klog.Errorf("Unable to fetch Machine object %s, Error: %+v", mc.Name, err)
+						return err
+					}
+					clone := mc.DeepCopy()
+					clone.Annotations[machinePriorityAnnotation] = "3"
+					_, err = m.machineClient.Machines(m.namespace).Update(ctx, clone, metav1.UpdateOptions{})
+					if err != nil {
+						select {
+						case <-ctx.Done():
+							return fmt.Errorf("could not reset priority annotation on machine %s; timed out, error: %+v", clone.Name, err)
+						case <-time.After(m.conflictRetryInterval):
+							klog.Warningf("could not reset priority annotation for machine %s; Error: %+v, will retry the operation", clone.Name, err)
+							continue
+						}
+					}
+					break
+				}
 			}
+			return nil
+		}()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-			if mclone.Annotations != nil {
-				if mclone.Annotations[machinePriorityAnnotation] == "1" {
-					klog.Infof("Machine %q priority is already set to 1, hence skipping the update", machine.Name)
+// prioritizeMachinesForDeletion prioritizes the targeted machines by updating their priority annotation to 1
+func (m *McmManager) prioritizeMachinesForDeletion(targetMachineRefs []*Ref, mdName string) ([]machineNameNodeNamePair, error) {
```

Comment:
why do you need `machineNameNodeNamePair`. The caller only uses the length of it

Reply:
It is used in a log.
```
	klog.V(2).Infof("MachineDeployment %s size decreased to %d , should remove following {machineRef, corresponding node} pairs %s ", commonMachineDeployment.Name, updatedReplicas, expectedToTerminateMachineNodePairs)
```